src/main/java/com/czertainly/core/service/cmp/message/builder/PkiMessageBuilder.java

```diff
@@ -0,0 +1,324 @@
+package com.czertainly.core.service.cmp.message.builder;
+
+import com.czertainly.core.api.cmp.error.CmpBaseException;
+import com.czertainly.core.api.cmp.error.CmpProcessingException;
+import com.czertainly.core.service.cmp.message.ConfigurationContext;
+import com.czertainly.core.service.cmp.message.PkiMessageDumper;
+import com.czertainly.core.service.cmp.message.protection.ProtectionStrategy;
+import com.czertainly.core.service.cmp.util.CertUtil;
+import com.czertainly.core.util.CertificateUtil;
+import org.bouncycastle.asn1.ASN1GeneralizedTime;
+import org.bouncycastle.asn1.ASN1OctetString;
+import org.bouncycastle.asn1.DEROctetString;
+import org.bouncycastle.asn1.cmp.*;
+import org.bouncycastle.asn1.crmf.CertReqMessages;
+import org.bouncycastle.asn1.crmf.CertReqMsg;
+import org.bouncycastle.asn1.crmf.CertRequest;
+import org.bouncycastle.asn1.x500.RDN;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x509.Certificate;
+import org.bouncycastle.asn1.x509.GeneralName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static com.czertainly.core.service.cmp.util.NullUtil.*;
+
+/**
+ * Builder pattern implementation to create specific types of {@link PKIMessage}.
+ */
+public class PkiMessageBuilder {

```

Comment:
Why we do not use implementation of `PKIMessage` from BouncyCastle and we need to have our own?

Reply:
please NEED discussion - @3keyroman or @lubomirw